libutils/glob_lib.c

```diff
@@ -0,0 +1,516 @@
+    ExpandBraces(pattern, expanded);
+    FREE_AND_NULL(pattern);
+
+    bool any_matches = false;
+    StringSetIterator iter = StringSetIteratorInit(expanded);
+    while ((pattern = StringSetIteratorNext(&iter)) != NULL)
+    {
+        char *const regex = TranslateGlob(pattern);
+        any_matches |= StringMatchFull(regex, filename);
+        free(regex);
+    }
+
+    StringSetDestroy(expanded);
+    free(filename);
+
+    return any_matches;
+}
+
+/**
+ * Returns true if basename represents current or parent directory (i.e., '.'
+ * or '..').
+ */
+static inline bool BasenameIsParentOrCurrent(const char *const path)
+{
+    assert(path != NULL);
+
+    char b_copy[strlen(path) + 1];
+    strcpy(b_copy, path);
+    const char *const b_name = basename(b_copy);
+    return StringEqual(b_name, ".") || StringEqual(b_name, "..");
+}
+
+/**
+ * Strategically walk through filesystem looking for files matching pattern.
+ */
+static void WalkPattern(
+    const char *const root,      // Currently matched subpath
+    const Seq *const components, // Glob components
+    const size_t index,          // Index of next (unmatched) glob component
+    Seq *const matches)          // Currently found matches
+{
+    assert(root != NULL);
+    assert(components != NULL);
+    assert(matches != NULL);
+
+    const size_t num_components = SeqLength(components);
+    assert(num_components > 0);
+    if (index >= num_components)
+    {
+        /* We have found a match, because we have recursively matched every
+         * glob component to a file or directory.
+         */
+        char *const match = NormalizePath(root);
+        SeqAppend(matches, match);
+        return;
+    }
+
+    const char *const component = SeqAt(components, index);
+    if (StringEqual(component, "."))
+    {
+        // We already know that this directory exists.
+        WalkPattern(root, components, index + 1, matches);
+    }
+
+    char *pattern =
+        (StringEndsWith(root, FILE_SEPARATOR_STR))
+            ? StringConcatenate(2, root, component)
+            : StringConcatenate(3, root, FILE_SEPARATOR_STR, component);
+
+    Seq *const files = ListDir(root, NULL);
+    const size_t num_files = (files == NULL) ? 0 : SeqLength(files);
+
+    for (size_t i = 0; i < num_files; i++)
+    {
+        const char *filename = SeqAt(files, i);
+        if (BasenameIsParentOrCurrent(filename))
+        {
+            // Iterating '.' & '..' directories would cause infinite recursion.
+        }
+        else if (GlobMatch(pattern, filename))
+        {
+            WalkPattern(filename, components, index + 1, matches);
+        }
+    }
+
+    SeqDestroy(files);
+    free(pattern);
+}
+
+Seq *GlobFind(const char *const pattern)
+{
+    assert(pattern != NULL);
+
+    Seq *const matches = SeqNew(1, free);
+
+    struct stat sb;
+    if (stat(pattern, &sb) == 0)
+    {
+        // The pattern itself is a match. There is no need to walk it.
```

Comment:
This doesn't always have to be the case, does it? If `pattern` contains special characters with some meaning in the pattern.